File: MiPrimeraAplicacion/app/src/main/java/com/micky/miprimeraaplicacion/WebServicesActivity.java
package com.micky.miprimeraaplicacion;

import android.os.Bundle;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import androidx.activity.EdgeToEdge;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.graphics.Insets;
import androidx.core.view.ViewCompat;
import androidx.core.view.WindowInsetsCompat;

import com.micky.miprimeraaplicacion.backend.WSAlbums;
import com.micky.miprimeraaplicacion.backend.WSPhotos;
import com.micky.miprimeraaplicacion.backend.WSUsers;
import com.micky.miprimeraaplicacion.backend.WebServiceAPI;
import com.squareup.picasso.Picasso;

import java.util.ArrayList;
import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class WebServicesActivity extends AppCompatActivity {
    private TextView textViewUserInfo;
    private TextView textViewPhotosInfo;

    private ImageView imageView;
    List<WSAlbums> listData = new ArrayList<>();
    List<WSUsers> listDataUsers = new ArrayList<>();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        EdgeToEdge.enable(this);
        setContentView(R.layout.activity_web_services);
        ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.main), (v, insets) -> {
            Insets systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars());
            v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom);
            return insets;
        });

        textViewUserInfo = findViewById(R.id.textViewUserInfo);
        imageView = findViewById(R.id.imageView);
        // textViewPhotosInfo= findViewById(R.id.textViewPhotosInfo);
        //getData();
        getDataUsers();
        getDataphotos();
    }
    public void getData(){
        String URL_WS = "https://jsonplaceholder.typicode.com/";
        final Retrofit retrofit = new Retrofit.Builder()
                .baseUrl(URL_WS)
                .addConverterFactory(GsonConverterFactory.create())
                .build();
        WebServiceAPI webServiceApi = retrofit.create(WebServiceAPI.class);
        final Call<List<WSAlbums>> call = webServiceApi.getAlbums();
        call.enqueue(new Callback<List<WSAlbums>>() {
            @Override
            public void onResponse(Call<List<WSAlbums>> call, Response<List<WSAlbums>> response) {
                if(response.isSuccessful()){
                    listData = response.body();
                    Toast.makeText(WebServicesActivity.this, listData.get(0).getTitle(), Toast.LENGTH_SHORT).show();
                }else{
                    Toast.makeText(WebServicesActivity.this, "Algo paso", Toast.LENGTH_SHORT).show();
                }
            }

            @Override
            public void onFailure(Call<List<WSAlbums>> call, Throwable throwable) {
                Toast.makeText(WebServicesActivity.this, "REVISE EL SERVICIO WEB DE INTERNET", Toast.LENGTH_SHORT).show();
            }
        });
    }

    public void getDataUsers(){
        String URL_WS = "https://jsonplaceholder.typicode.com/";
        final Retrofit retrofit = new Retrofit.Builder()
                .baseUrl(URL_WS)
                .addConverterFactory(GsonConverterFactory.create())
                .build();
        WebServiceAPI webServiceApi = retrofit.create(WebServiceAPI.class);
        final Call<List<WSUsers>> call = webServiceApi.getUsers();
        call.enqueue(new Callback<List<WSUsers>>() {
            @Override
            public void onResponse(Call<List<WSUsers>> call, Response<List<WSUsers>> response) {
                if(response.isSuccessful()){
                    listDataUsers = response.body();
                    StringBuilder userInfoBuilder = new StringBuilder();
                    //Toast.makeText(WebServicesActivity.this, listDataUsers.get(0).getAddress().getCity(), Toast.LENGTH_SHORT).show();

                    // Recorrer la lista de usuarios y construir la cadena con los detalles de cada usuario
                    for (WSUsers user : listDataUsers) {
                        userInfoBuilder.append("---------------------------------------\n");
                        userInfoBuilder.append("ID: ").append(user.getId()).append("\n");
                        userInfoBuilder.append("Name: ").append(user.getName()).append("\n");
                        userInfoBuilder.append("Username: ").append(user.getUsername()).append("\n");
                        userInfoBuilder.append("Email: ").append(user.getEmail()).append("\n");
                        userInfoBuilder.append("Phone: ").append(user.getPhone()).append("\n");
                        userInfoBuilder.append("Website: ").append(user.getWebsite()).append("\n");
                        userInfoBuilder.append("Address: ").append("\n");
                        userInfoBuilder.append("   Street: ").append(user.getAddress().getStreet()).append("\n");
                        userInfoBuilder.append("   Suite: ").append(user.getAddress().getSuite()).append("\n");
                        userInfoBuilder.append("   City: ").append(user.getAddress().getCity()).append("\n");
                        userInfoBuilder.append("   Zipcode: ").append(user.getAddress().getZipcode()).append("\n");
                        userInfoBuilder.append("   Geo: ").append("\n");
                        userInfoBuilder.append("       Lat: ").append(user.getAddress().getGeo().getLat()).append("\n");
                        userInfoBuilder.append("       Lng: ").append(user.getAddress().getGeo().getLng()).append("\n");
                        userInfoBuilder.append("Company: ").append("\n");
                        userInfoBuilder.append("   Name: ").append(user.getCompany().getName()).append("\n");
                        userInfoBuilder.append("   Catch Phrase: ").append(user.getCompany().getCatchPhrase()).append("\n");
                        userInfoBuilder.append("   Bs: ").append(user.getCompany().getBs()).append("\n\n");
                    }

                    // Mostrar la información en el TextView
                    textViewUserInfo.setText(userInfoBuilder.toString());
                }else{
                    Toast.makeText(WebServicesActivity.this, "Algo paso", Toast.LENGTH_SHORT).show();
                }
            }

            @Override
            public void onFailure(Call<List<WSUsers>> call, Throwable throwable) {
                Toast.makeText(WebServicesActivity.this, "REVISE EL SERVICIO WEB DE INTERNET", Toast.LENGTH_SHORT).show();
            }
        });
    }

    public void getDataphotos(){
        String URL_WS = "https://jsonplaceholder.typicode.com/";
        final Retrofit retrofit = new Retrofit.Builder()
                .baseUrl(URL_WS)
                .addConverterFactory(GsonConverterFactory.create())
                .build();
        WebServiceAPI webServiceApi = retrofit.create(WebServiceAPI.class);
        final Call<List<WSPhotos>> call = webServiceApi.getPhotos();

        call.enqueue(new Callback<List<WSPhotos>>() {
            @Override
            public void onResponse(Call<List<WSPhotos>> call, Response<List<WSPhotos>> response) {
                if (response.isSuccessful()) {
                    List<WSPhotos> listDataPhotos = response.body();
                    if (listDataPhotos != null && !listDataPhotos.isEmpty()) {
                        WSPhotos photo = listDataPhotos.get(10); // Obtener la foto en la posición 0
                        String photoUrl = photo.getUrl(); // Obtener la URL de la foto

                        // Cargar la imagen de la URL en un ImageView usando Picasso
                        Picasso.get().load(photoUrl).into(imageView);
                        // Cambia "imageView" por el ID de tu ImageView
                    } else {
                        Toast.makeText(WebServicesActivity.this, "La lista de fotos está vacía", Toast.LENGTH_SHORT).show();
                    }
                } else {
                    Toast.makeText(WebServicesActivity.this, "Algo pasó", Toast.LENGTH_SHORT).show();
                }
            }

            @Override
            public void onFailure(Call<List<WSPhotos>> call, Throwable throwable) {
                Toast.makeText(WebServicesActivity.this, "REVISE EL SERVICIO WEB DE INTERNET", Toast.LENGTH_SHORT).show();
            }
        });

    }
}

File: PrimerKotlin/app/src/main/java/com/micky/primerkotlin/WebServicesActivity.kt
package com.micky.primerkotlin

//import java.util.function.Function;

import android.os.Bundle
import android.util.Log
import android.widget.Button
import android.widget.ImageView
import android.widget.TextView
import androidx.activity.*
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import com.google.gson.GsonBuilder
import com.micky.primerkotlin.backend.WSDirecciones
import com.micky.primerkotlin.backend.WSInformaciones
import com.micky.primerkotlin.backend.WSPersonas
import com.micky.primerkotlin.backend.WSProfesiones
import com.micky.primerkotlin.backend.WebServiceAPI
import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import io.reactivex.observers.DisposableObserver
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import retrofit2.converter.gson.GsonConverterFactory

class WebServicesActivity : AppCompatActivity() {
    private val textViewUserInfo: TextView? = null
    private val textViewPhotosInfo: TextView? = null

    private val imageView: ImageView? = null
    private val textViewResultado:     TextView? = null

    var buttonTiendaOnline2: Button? = null
    var listaPersonas: List<WSPersonas> = java.util.ArrayList<WSPersonas>()
    var listaProfesiones: List<WSProfesiones> = java.util.ArrayList<WSProfesiones>()
    var listaInformaciones: List<WSInformaciones> = java.util.ArrayList<WSInformaciones>()
    var listaDirecciones: List<WSDirecciones> = java.util.ArrayList<WSDirecciones>()
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        enableEdgeToEdge()
        setContentView(R.layout.activity_web_services)
        ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.main)) { v, insets ->
            val systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars())
            v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom)
            insets
        }
        getPersonas()
    }

    fun getPersonas() {
        val URL_WS = "http://192.168.0.1/webservice/    "
        val gson = GsonBuilder().setLenient().create()

        val interceptor = HttpLoggingInterceptor()
        interceptor.setLevel(HttpLoggingInterceptor.Level.BODY)
        val client: OkHttpClient = OkHttpClient.Builder().addInterceptor(interceptor).build()

        val retrofit2 = Retrofit.Builder()
            .baseUrl(URL_WS)
            .client(client)
            .addCallAdapterFactory(RxJava2CallAdapterFactory.createAsync())
            .addConverterFactory(GsonConverterFactory.create(gson))
            .build()

        val webServiceApi: WebServiceAPI = retrofit2.create(WebServiceAPI::class.java)
        val requests: MutableList<Observable<*>> = ArrayList()


        requests.add(webServiceApi.getPersonasObservable())
        requests.add(webServiceApi.getProfesionesObservable())
        requests.add(webServiceApi.getInformacionesObservable())
        requests.add(webServiceApi.getDireccionesObservable())

        val obs: Disposable = Observable.zip<Any, Any>(
            requests
        ) { objects ->
            listaPersonas = objects[0] as List<WSPersonas>
            listaProfesiones = objects[1] as List<WSProfesiones>
            listaDirecciones = objects[2] as List<WSDirecciones>
            listaInformaciones = objects[3] as List<WSInformaciones>
            Any()
        }
            .observeOn(AndroidSchedulers.mainThread())
            .subscribeWith(object : DisposableObserver<Any?>() {
                override fun onNext(o: Any) {
                    // Método requerido pero no necesitamos hacer nada aquí
                }

                override fun onError(e: Throwable) {
                    Log.d("Datos", "OnError ")
                }

                override fun onComplete() {
                    Log.d("Datos", "Finalizado con éxito onComplete ")

                    val persona: WSPersonas = listaPersonas.get(2)
                    val profesion: WSProfesiones = listaProfesiones.get(2)
                    val informacion: WSInformaciones = listaInformaciones.get(2)
                    val direccion: WSDirecciones = listaDirecciones.get(2)

                    val dato = StringBuilder()
                    dato.append("PERSONA: \n")
                    dato.append("ID: ").append(persona.getId()).append("\n")
                    dato.append("Nombre: ").append(persona.getNombre()).append("\n")
                    dato.append("Edad: ").append(persona.getApellidos()).append("\n")
                    dato.append("Email: ").append(persona.getCi()).append("\n")

                    dato.append("PROFESION: \n")
                    dato.append("ID: ").append(profesion.getId()).append("\n")
                    dato.append("Carrera: ").append(profesion.getCarrera()).append("\n")

                    dato.append("INFORMACION: \n")
                    dato.append("ID: ").append(informacion.getId()).append("\n")
                    dato.append("Correo: ").append(informacion.getCorreo()).append("\n")
                    dato.append("Celular: ").append(informacion.getCelular()).append("\n")
                        .append("\n\n")

                    dato.append("DIRECCION: \n")
                    dato.append("ID: ").append(direccion.getId()).append("\n")
                    dato.append("Calle: ").append(direccion.getCalle()).append("\n")
                    dato.append("Numero: ").append(direccion.getNumero()).append("\n")

                    textViewResultado?.setText(dato.toString())
                }
            })
    }


}